### PR TITLE
Add a `java(\\d+)` flag

### PR DESCRIPTION
**Pull request by jrhee17**
Motivation:

Some modules may need to target a java version higher than other modules to compile correctly.
This PR proposes a `java(\\d+)` flag which sets the target java version for compilation.

Modifications:

- Remove special logic for `:examples` in `build.gradle`
- Set the `-release` flag in `core/build.gradle` for `buildJdkVersion >= 9` to avoid compilation errors.
    - Without this change, the actual error message might be shadowed (i.e. invalid flag exception instead of invalid class version exception)
- Introduce an `extractTargetJavaVersion` method which extracts the target version from a `java(\\d+)` flag.
    - The default is set to 8. If a lower version than the default is specified an exception is thrown. (e.g. `java3`)
- Set the compatibility to `targetJavaVersion`, and skip tests if `testJavaVersion` is lower than the `targetJavaVersion`
- Imply the `java` flag if `java(\\d+)` flag is added

Result:

- Users can support modules with different minimum java versions easily.

Note: If multiple `java(\\d+)` flags are specified, only the first one will have effect.

### CODE REVIEW

File: gradle/scripts/lib/java.gradle [hunk 1]
```diff
@@ -34,6 +36,31 @@ configure(rootProject) {
...
+    def flags = project.ext.flags
+    for (def flag: flags) {
+        def matcher = pattern.matcher(flag)
+        if (!matcher.matches()) {
+            continue
+        }
+        def version = Integer.parseInt(matcher.group(1))
+        if (version < defaultVersion) {
+            throw new IllegalArgumentException("The minimum java version ${version} specified " +
+                    "by '${flag}' cannot be smaller than s${defaultVersion}")
```
Comment (minwoox) [inline gradle/scripts/lib/java.gradle:57]:
```suggestion
                    "by '${flag}' cannot be smaller than ${defaultVersion}.")
```

File: gradle/scripts/lib/java.gradle [hunk 2]
```diff
@@ -34,6 +36,31 @@ configure(rootProject) {
     apply plugin: 'idea'
 }
 
+
+/**
+ * Checks each flag of the specified project for flags of format "java(\\d+)".
+ * If such a flag exists, the minimum jre version is extracted and returned.
```
Comment (minwoox) [inline gradle/scripts/lib/java.gradle:42]:
```suggestion
 * If such a flag exists, the minimum JRE version is extracted and returned.
```

File: gradle/scripts/README.md [hunk 3]
```diff
@@ -639,6 +639,32 @@ for more information.
    }
    ```
 
+## Setting a target version with the `java(\\d+)` flag.
+
+By default, setting the `java` flag compiles a module targeting minimum java compatibility with java 8.
```
Comment (minwoox) [inline gradle/scripts/README.md:644]:
```suggestion
By default, setting the `java` flag compiles a module targeting minimum Java compatibility with Java 8.
```
Please also fix the word in other sentences as well.

File: gradle/scripts/lib/java.gradle [hunk 4]
```diff
@@ -91,8 +122,12 @@ configure(projectsWithFlags('java')) {
 
     // Set the sensible compiler options.
     tasks.withType(JavaCompile) {
-        sourceCompatibility = project.findProperty('javaSourceCompatibility') ?: '1.8'
-        targetCompatibility = project.findProperty('javaTargetCompatibility') ?: '1.8'
+        sourceCompatibility = project.findProperty('javaSourceCompatibility') ?: "${targetJavaVersion}"
```
Comment (minwoox) [inline gradle/scripts/lib/java.gradle:125]: Does this work when it's JDK 1.8?
previously it was `1.8` but now it's `8`.

**Review by minwoox — COMMENTED** [review]
Left minor comments. 😉

File: gradle/scripts/lib/java.gradle [hunk 5]
```diff
@@ -125,6 +160,11 @@ configure(projectsWithFlags('java')) {
                 languageVersion = JavaLanguageVersion.of(rootProject.ext.testJavaVersion)
             }
         }
+
+        // disable tests for projects which target a higher java version
+        if (testJavaVersion < targetJavaVersion) {
```
Comment (minwoox) [inline gradle/scripts/lib/java.gradle:165]: nit: For consistency: `rootProject.ext.testJavaVersion` 😄

File: gradle/scripts/lib/java.gradle (hunk 4 above)
Comment (jrhee17) [reply gradle/scripts/lib/java.gradle:125]: I believe the following method is called before usage internally

https://github.com/gradle/gradle/blob/5ec3f672ed600a86280be490395d70b7bc634862/subprojects/base-services/src/main/java/org/gradle/api/JavaVersion.java#L139

I also found unit tests supporting this
https://github.com/gradle/gradle/blob/585397da80852998840f19638057c90df2c8302a/subprojects/java-compiler-plugin/build.gradle.kts#L10

Comment (minwoox) [reply gradle/scripts/lib/java.gradle:125]: I didn't know that Gradle covers both cases. Thanks! 😄

Comment (jrhee17) [reply gradle/scripts/lib/java.gradle:125]: Let me just pass a `JavaVersion` instead of a string to make the intention clearer 😅

**Review by ikhoon — COMMENTED** [review]
This is super useful!

File: gradle/scripts/README.md (hunk 3 above)
Comment (jrhee17) [reply gradle/scripts/README.md]: Thanks, fixed for all public documentation

File: gradle/scripts/lib/java.gradle [hunk 6]
```diff
@@ -50,6 +77,10 @@ configure(projectsWithFlags('java')) {
         delete project.ext.genSrcDir
     }
 
+    // parse flags for the target java version with the default as java 8.
+    // an exception will be thrown if a version <8 is specified in the flags.
+    def targetJavaVersion = extractTargetJavaVersion(project, 8)
```
[the remark this answers is not in the document]
Comment (jrhee17) [reply gradle/scripts/lib/java.gradle]: Good idea, fixed to use `javaTargetCompatibility` as the default target version 👍

File: gradle/scripts/lib/java.gradle (hunk 5 above)
Comment (jrhee17) [reply gradle/scripts/lib/java.gradle]: Both done 😄

**Review by ikhoon — APPROVED** [review]
This would be super useful. Thanks, @jrhee17! 💯💯

**Review by minwoox — APPROVED** [review]
Thanks! 😄

**Pull request merged**